query/src/sql/planner/binder/select.rs

```diff
@@ -176,11 +178,29 @@ impl<'a> Binder {
         }
     }
 
+    #[async_recursion]
     pub(crate) async fn bind_query(
         &mut self,
-        bind_context: &BindContext,
+        bind_context: &mut BindContext,
         query: &Query<'_>,
     ) -> Result<(SExpr, BindContext)> {
+        if let Some(with) = &query.with {
+            for cte in with.ctes.iter() {
+                let table_name = cte.alias.name.name.clone();
+                if bind_context.ctes_map.contains_key(&table_name) {
+                    return Err(ErrorCode::SemanticError(format!(
+                        "duplicate cte {table_name}"
+                    )));
+                }
+                let (s_expr, cte_bind_context) = self.bind_query(bind_context, &cte.query).await?;
+                let cte_info = CteInfo {
+                    columns_alias: cte.alias.columns.iter().map(|c| c.name.clone()).collect(),
+                    s_expr,
+                    bind_context: cte_bind_context.clone(),
+                };
+                bind_context.ctes_map.insert(table_name, cte_info);
```

Comment:
It would be better to change ctes_map to rwlock and add a function for the map operation? Make all bind_context to immutable.

Reply:
Make sense